File: pipesock/pipesock.ts
import path from 'path'
import net from 'net'
import { pushable } from 'it-pushable'
import { pipe } from 'it-pipe'
import {
  StreamConn,
  buildPushableSink,
  combineUint8ArrayListTransform,
} from 'starpc'

/**
 * Builds a pipe name for IPC communication
 * @param rootDir - The root directory where the pipe will be created (used for unix sockets)
 * @param pipeUuid - Unique identifier for the pipe
 * @returns The platform-specific pipe path
 */
export function buildPipeName(rootDir: string, pipeUuid: string): string {
  if (process.platform === 'win32') {
    return `\\\\.\\pipe\\aptre\\${pipeUuid}`
  } else {
    // Create absolute path for the socket
    const absolutePath = path.join(rootDir, `.pipe-${pipeUuid}`)
    try {
      // Get relative path from current working directory if possible
      // Use whichever is shorter (Unix socket paths are limited to ~104 chars)
      const relPath = path.relative(process.cwd(), absolutePath)
      if (relPath.length < absolutePath.length) {
        return relPath
      }
      return absolutePath
    } catch {
      // If we can't get CWD (e.g., it was deleted), use absolute path
      return absolutePath
    }
  }
}

/**
 * Socket connection with pushable streams
 */
export interface SocketConnection {
  /** Socket instance */
  socket: net.Socket
  /** Stream for sending data to the socket */
  socketTx: ReturnType<typeof pushable<Uint8Array>>
  /** Stream for receiving data from the socket */
  socketRx: ReturnType<typeof pushable<Uint8Array>>
  /** SRPC stream connection */
  streamConn: StreamConn
}

/**
 * Creates a socket connection with pushable streams
 * @param socket - The socket instance
 * @param streamConn - The SRPC stream connection
 * @returns The socket connection with pushable streams
 */
export function createSocketConnection(
  socket: net.Socket,
  streamConn: StreamConn,
): SocketConnection {
  // Set up bidirectional communication
  const socketTx = pushable<Uint8Array>({ objectMode: true })
  const socketRx = pushable<Uint8Array>({ objectMode: true })

  // Pipe data between socket and SRPC
  pipe(
    socketRx,
    streamConn,
    combineUint8ArrayListTransform(),
    buildPushableSink<Uint8Array>(socketTx),
  ).catch((err) => {
    console.error(`[pipesock] pipe error: ${err}`)
    streamConn.close(err)
  })

  // Handle socket data
  socket.on('data', (data) => {
    if (typeof data === 'string') {
      throw new Error('unexpected string data from socket')
    }
    socketRx.push(data)
  })

  // Handle socket close
  socket.on('end', () => {
    console.log('[pipesock] socket closed')
    socketRx.end()
    streamConn.close()
  })

  // Handle socket errors
  socket.on('error', (err) => {
    console.error(`[pipesock] socket error: ${err}`)
    socketRx.end(err)
    streamConn.close(err)
  })

  return { socket, socketTx, socketRx, streamConn }
}

/**
 * Starts sending data from the socketTx stream to the socket
 * @param connection - The socket connection
 * @returns A promise that resolves when the sending is complete
 */
export async function startSocketSender(
  connection: SocketConnection,
): Promise<void> {
  try {
    for await (const data of connection.socketTx) {
      connection.socket.write(data)
    }
  } catch (err) {
    console.error(`[pipesock] send error: ${err}`)
    connection.streamConn.close(err as Error)
  }
}

/**
 * Connects to a pipe and sets up bidirectional communication
 * @param ipcPath - The path to the pipe
 * @param streamConn - The SRPC stream connection
 * @param onConnect - Callback function called when the connection is established
 * @returns The socket instance
 */
export function connectToPipe(
  ipcPath: string,
  streamConn: StreamConn,
  onConnect?: (connection: SocketConnection) => void,
): net.Socket {
  const socket = net.connect(ipcPath, () => {
    console.log(`[pipesock] connected to pipe: ${ipcPath}`)

    const connection = createSocketConnection(socket, streamConn)

    // Start sending data to socket
    startSocketSender(connection)

    // Call the onConnect callback if provided
    if (onConnect) {
      onConnect(connection)
    }
  })

  // Handle connection errors
  socket.on('error', (err) => {
    console.error(`[pipesock] connection error: ${err}`)
  })

  return socket
}

File: pipesock/pipesock.go
//go:build !windows

package pipesock

import (
	"context"
	"net"
	"os"
	"path/filepath"

	"github.com/pkg/errors"
	"github.com/sirupsen/logrus"
)

// resolvePipePath builds the unix socket path for a pipe.
// The kernel bounds sun_path, so an over-long path is rejected here by name
// rather than at bind time as an opaque invalid argument.
func resolvePipePath(le *logrus.Entry, rootDir, pipeUuid string) (string, error) {
	// Create absolute path for the socket
	absolutePipePath := filepath.Join(rootDir, ".pipe-"+pipeUuid)

	// Get current working directory
	cwd, err := os.Getwd()
	if err != nil {
		// If we can't get CWD (e.g., it was deleted), use absolute path
		le.WithError(err).Debug("could not get cwd, using absolute pipe path")
		cwd = ""
	}

	// Get relative path from current working directory if possible
	// Use whichever is shorter, since sun_path is small.
	pipePath := absolutePipePath
	if cwd != "" {
		relPath, err := filepath.Rel(cwd, absolutePipePath)
		if err == nil && len(relPath) < len(pipePath) {
			pipePath = relPath
		}
	}

	if len(pipePath) > maxPipePathLen {
		return "", errors.Errorf(
			"unix socket path is %d bytes and this platform binds at most %d: %s",
			len(pipePath), maxPipePathLen, pipePath,
		)
	}
	return pipePath, nil
}

// BuildPipeListener builds the pipe listener.
// The rootDir is used for unix sockets if this is a linux system.
// The pipeUuid is used for the socket path OR the Windows Pipe Name.
// The pipeUuid should be unique to the local device and pipe.
func BuildPipeListener(le *logrus.Entry, rootDir, pipeUuid string) (net.Listener, error) {
	// Ensure the parent directory exists
	pipeDir := filepath.Dir(filepath.Join(rootDir, ".pipe-"+pipeUuid))
	if err := os.MkdirAll(pipeDir, 0o755); err != nil {
		return nil, errors.Wrap(err, "create pipe directory")
	}

	pipePath, err := resolvePipePath(le, rootDir, pipeUuid)
	if err != nil {
		return nil, err
	}

	// remove old pipe file, if exists
	if _, err := os.Stat(pipePath); !os.IsNotExist(err) {
		if err := os.Remove(pipePath); err != nil {
			return nil, errors.Wrap(err, "remove old pipe file")
		}
	}

	addr := &net.UnixAddr{
		Net:  "unix",
		Name: pipePath,
	}
	le.Debugf("listening on unix socket: %s", addr.String())
	return net.ListenUnix("unix", addr)
}

// DialPipeListener connects to the pipe listener in the directory.
func DialPipeListener(ctx context.Context, le *logrus.Entry, rootDir, pipeUuid string) (net.Conn, error) {
	pipePath, err := resolvePipePath(le, rootDir, pipeUuid)
	if err != nil {
		return nil, err
	}

	addr := &net.UnixAddr{
		Net:  "unix",
		Name: pipePath,
	}
	le.Debugf("connecting to unix socket: %s", addr.String())
	dialer := net.Dialer{}
	return dialer.DialContext(ctx, "unix", addr.String())
}
